src/settings/toml/tests/deploy_target.rs

```diff
@@ -335,3 +335,540 @@ fn it_errors_on_deploy_target_routes_and_workers_dev_true() {
+
+    assert!(actual_deploy_target.is_err());
+
+    // when route is empty, error
+    let patterns = [""];
+    let env_config = EnvConfig::zoned_multi_route(zone_id, patterns.to_vec());
+
+    let script = "top_level_empty_env_zoned_multi_route_empty";
+    let test_toml = WranglerToml::webpack_with_env(script, env_name, env_config);
+    let toml_string = toml::to_string(&test_toml).unwrap();
+    let manifest = Manifest::new_from_string(toml_string).unwrap();
+
+    let actual_deploy_target = manifest.deploy_target(Some(env_name));
+
+    assert!(actual_deploy_target.is_err());
+
+    // when zone id is missing, error
+    let patterns = ["hostname.tld/*"];
+    let env_config = EnvConfig::zoned_multi_route("", patterns.to_vec());
+
+    let script = "top_level_empty_env_zoned_multi_route_no_zone_id";
+    let test_toml = WranglerToml::webpack_with_env(script, env_name, env_config);
+    let toml_string = toml::to_string(&test_toml).unwrap();
+    let manifest = Manifest::new_from_string(toml_string).unwrap();
+
+    let actual_deploy_target = manifest.deploy_target(Some(env_name));
+
+    assert!(actual_deploy_target.is_err());
+
+    // when zone id is present, all good
+    let patterns = ["hostname.tld/*"];
+    let env_config = EnvConfig::zoned_multi_route(zone_id, patterns.to_vec());
+
+    let script = "top_level_empty_env_zoned_multi_route_no_zone_id";
+    let test_toml = WranglerToml::webpack_with_env(script, env_name, env_config);
+    let toml_string = toml::to_string(&test_toml).unwrap();
+    let manifest = Manifest::new_from_string(toml_string).unwrap();
+
+    let actual_deploy_target = manifest.deploy_target(Some(env_name)).unwrap();
+
+    let expected_name = manifest.worker_name(Some(env_name));
+
+    let expected_routes = patterns
+        .iter()
+        .map(|p| Route {
+            script: Some(expected_name.to_string()),
+            pattern: p.to_string(),
+            id: None,
+        })
+        .collect();
+
+    let expected_deploy_target = DeployTarget::Zoned(Zoned {
+        zone_id: zone_id.to_string(),
+        routes: expected_routes,
+    });
+
+    assert_eq!(actual_deploy_target, expected_deploy_target);
+}
+
+#[test]
+fn when_top_level_zoneless_env_empty() {
+    let script = "top_level_zoneless_env_empty";
+    let env_name = "test";
+    let env_config = EnvConfig::default();
+
+    let test_toml = WranglerToml::webpack_zoneless_with_env(script, true, env_name, env_config);
+    let toml_string = toml::to_string(&test_toml).unwrap();
+    let manifest = Manifest::new_from_string(toml_string).unwrap();
+
+    let actual_deploy_target = manifest.deploy_target(Some(env_name)).unwrap();
+
+    assert_eq!(actual_deploy_target, DeployTarget::Zoneless);
+}
+
+#[test]
+fn when_top_level_zoneless_env_zoneless() {
+    let env_name = "test";
+
+    // TODO: SHOULD THIS TEST PASS?
+    // when env.workers_dev = false
+    let env_config = EnvConfig::zoneless(false);
```

Comment:
this should require a route and zone_id to pass, right? For now, this test as written should fail.

Reply:
the assert is that this errors.

Reply:
the todo says this should pass, though. if it does pass, I think it needs to meet the conditions I mentioned in this comment.

Reply:
by "test should pass" i mean the assert should be true; in this case, should error.